src/twister2/device/native_simulator_adapter.py

```diff
@@ -23,78 +29,103 @@ def __init__(self, twister_config, hardware_map: HardwareMap | None = None, **kw
+        assert isinstance(command, (list, tuple, set))  # to avoid stupid and  difficult to debug mistakes
+        # we are using asyncio to run subprocess to be able to read from stdout
+        # without blocking while loop (readline with timeout)
+        self._process = await asyncio.create_subprocess_exec(
+            *command,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.STDOUT,
+            env=self.env
+        )
+        logger.debug('Started subprocess with PID %s', self._process.pid)
+        end_time = time.time() + timeout
+        while not self._stop_job and not self._process.stdout.at_eof():
+            try:
+                line = await asyncio.wait_for(self._process.stdout.readline(), timeout=0.1)
+            except asyncio.TimeoutError:
+                pass
             else:
-                if stderr := self._process.stderr.read():
-                    logger.error(stderr.decode())
-                raise TwisterFlashException(f'Could not run simulator with PID {self._process.pid}')
-
-    def _collect_process_output(self, process: subprocess.Popen) -> threading.Thread:
-        """Create Thread which saves a process output to a file."""
-        def _read():
-            with process.stdout:
-                for line in iter(process.stdout.readline, b''):
-                    self.queue.put(line.decode().strip())
-
-        return threading.Thread(target=_read, daemon=True)
-
-    def _wait_and_terminate_process(self, process: subprocess.Popen, timeout: float) -> threading.Thread:
-        """Create Thread which kills a process after given time."""
-        def _waiting():
-            end_time = time.time() + timeout
-            while process.poll() is None and time.time() < end_time:
-                time.sleep(0.1)
-            if process.poll() is None:
-                process.kill()
-                logger.error('Process %s terminated after %s seconds', process.pid, timeout)
+                if line:
+                    self.queue.put(line.decode('utf-8').strip())
+            if time.time() > end_time:
                 self._process_ended_with_timeout = True
-        return threading.Thread(target=_waiting, daemon=True)
+                logger.debug(f'Finished process with PID {self._process.pid} after {timeout} seconds timeout')
+                break
+
+        self.queue.put(END_DATA)
+        self.queue.task_done()
+        return await self._process.wait()
+
+    def _task(self) -> None:
+        command: list[str] = self._get_command(self.build_dir)
+        logger.info('Running command: %s', command)
+        try:
+            return_code: int = asyncio.run(
+                self._run_command(command, timeout=self.timeout)
+            )
+        except subprocess.SubprocessError as e:
+            logger.error('Flashing failed due to subprocess error %s', e)
+            self._exc = TwisterFlashException(e.args)
```

Comment:
I wondering if this is good strategy to save only this error without rising them immediately somehow. Let's consider following scenario:
 1. Test was built for `natvie_posix` successfully.
 2. `zephyr.exe` is run in `_run_command` method, but this exe rises `SubprocessError` immediately.
 3. Error is saved without rising it, but log parser will wait for some output until timeout, and only then user will see, that something wrong happen during exe execution - it can take too much time and be frustrating.

Maybe here in `_task` method at the end try-except there should be added `finally` with putting into `queue` info about end of stream (in the same way as it is done at the end of `_run_command` method):
```python
finnaly:
    self.queue.put(END_DATA)
    self.queue.task_done()
```

Reply:
This is a problem with unhandled exception in thread. Pytest will print a warning: `warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))` which very unclear and may be confusing for a user. This implementation with raising an exception in thread.join() method was one I found. But adding this example with `finnaly` should resolve a problem with waiting for data in test execution. Maybe we could also put some message to the queue which can be parsed in test and displayed for a user e.g. Test failed due to an error in subprocess.